rust/agama-cli/src/logs.rs

```diff
@@ -0,0 +1,257 @@
+    "journalctl -u agama-auto",
+    "journalctl --dmesg",
+];
+
+const DEFAULT_PATHS: [&str; 14] = [
+    // logs
+    "/var/log/YaST2",
+    "/var/log/zypper.log",
+    "/var/log/zypper/history*",
+    "/var/log/zypper/pk_backend_zypp",
+    "/var/log/pbl.log",
+    "/var/log/linuxrc.log",
+    "/var/log/wickedd.log",
+    "/var/log/NetworkManager",
+    "/var/log/messages",
+    "/var/log/boot.msg",
+    "/var/log/udev.log",
+    // config
+    "/etc/install.inf",
+    "/etc/os-release",
+    "/linuxrc.config",
+];
+
+const DEFAULT_RESULT: &str = "/tmp/agama_logs";
+const DEFAULT_COMPRESSION: (&str, &str) = ("bzip2", "tar.bz2");
+const DEFAULT_TMP_DIR: &str = "agama-logs";
+
+// A wrapper around println which shows (or not) the text depending on the boolean variable
+fn showln(show: bool, text: &str) {
+    if !show {
+        return;
+    }
+
+    println!("{}", text);
+}
+
+// A wrapper around println which shows (or not) the text depending on the boolean variable
+fn show(show: bool, text: &str) {
+    if !show {
+        return;
+    }
+
+    print!("{}", text);
+}
+
+// Struct for log represented by a file
+struct LogPath {
+    // log source
+    src_path: String,
+
+    // directory where to collect logs
+    dst_path: PathBuf,
+}
+
+impl LogPath {
+    fn new(src: &str, dst: &Path) -> Self {
+        Self {
+            src_path: src.to_string(),
+            dst_path: dst.to_owned(),
+        }
+    }
+}
+
+// Struct for log created on demmand by a command
+struct LogCmd {
+    // command which stdout / stderr is logged
+    cmd: String,
+
+    // place where to collect logs
+    dst_path: PathBuf,
+}
+
+impl LogCmd {
+    fn new(cmd: &str, dst: &Path) -> Self {
+        Self {
+            cmd: cmd.to_string(),
+            dst_path: dst.to_owned(),
+        }
+    }
+}
+
+trait LogItem {
+    // definition of log source
+    fn from(&self) -> &String;
+
+    // definition of destination as path to a file
+    fn to(&self) -> PathBuf;
+
+    // performs whatever is needed to store logs from "from" at "to" path
+    fn store(&self) -> Result<(), Error>;
+}
+
+impl LogItem for LogPath {
+    fn from(&self) -> &String {
+        &self.src_path
+    }
+
+    fn to(&self) -> PathBuf {
+        // remove leading '/' if any from the path (reason see later)
+        let r_path = Path::new(self.src_path.as_str()).strip_prefix("/").unwrap();
+
+        // here is the reason, join overwrites the content if the joined path is absolute
+        self.dst_path.join(r_path)
+    }
+
+    fn store(&self) -> Result<(), Error> {
+        // for now keep directory structure close to the original
+        // e.g. what was in /etc will be in /<tmp dir>/etc/
+        fs::create_dir_all(self.to().parent().unwrap())?;
+
+        let options = CopyOptions::new();
+        // fs_extra's own Error doesn't implement From trait so ? operator is unusable
+        match copy_items(
+            &[self.src_path.as_str()],
+            self.to().parent().unwrap(),
```

Comment:
You are doing exactly the same in line 143. What about writing to a variable with a nice name?

Reply:
done